Sources/Basics/Observability.swift

```diff
@@ -0,0 +1,400 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2021 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+ */
+
+import TSCBasic
+import TSCUtility
+
+typealias TSCDiagnostic = TSCBasic.Diagnostic
+
+// this could become a struct when we remove the "errorsReported" pattern
+
+// designed after https://github.com/apple/swift-log
+// designed after https://github.com/apple/swift-metrics
+public class ObservabilitySystem {
+
+    // global
+
+    private static var _global = ObservabilitySystem(factory: NOOPFactory())
+    private static var bootstrapped = false
+    private static let lock = Lock()
+
+    public static func bootstrapGlobal(factory: ObservabilityFactory) {
+        Self.lock.withLock {
+            // FIXME: disabled for testing
+            //precondition(!Self.bootstrapped, "ObservabilitySystem can only bootstrapped once")
+            Self._global = .init(factory: factory)
+            Self.bootstrapped = true
+        }
+    }
+
+    @available(*, deprecated, message: "this pattern is deprecated, transition to error handling instead")
```

Comment:
There is one use case I can see for this, and that is for code that wants to make sure that there is at least a fallback error reported if none have been reported so far.  That situation is of course usually the result of bad pattern to begin with — these kinds of "catch-all" errors have caused problems in the past.  But for example when shelling out to a tool and interpreting its errors, it's useful to know whether any errors were have been emitted when deciding whether to emit a "tool failed" error.  Arguably that's a special enough case that it should be handled by the custom code that is interpreting the tool's diagnostics in the first place — I'm probably overthinking this case, but wanted to mention it.